File: src/main/java/work/xeltica/craft/core/handlers/WorldHandler.java
package work.xeltica.craft.core.handlers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

import io.papermc.paper.event.block.BlockPreDispenseEvent;
import org.bukkit.Bukkit;
import org.bukkit.Difficulty;
import org.bukkit.GameMode;
import org.bukkit.GameRule;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.SoundCategory;
import org.bukkit.block.Dispenser;
import org.bukkit.entity.Entity;
import org.bukkit.entity.GlowItemFrame;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.player.PlayerAdvancementDoneEvent;
import org.bukkit.event.player.PlayerChangedWorldEvent;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.inventory.BlockInventoryHolder;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.inventory.ShapelessRecipe;
import org.bukkit.material.Directional;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.title.Title;

import work.xeltica.craft.core.XCorePlugin;
import work.xeltica.craft.core.models.CraftRecipe;
import work.xeltica.craft.core.modules.hint.Hint;
import work.xeltica.craft.core.models.PlayerDataKey;
import work.xeltica.craft.core.modules.hint.HintModule;
import work.xeltica.craft.core.stores.PlayerStore;
import work.xeltica.craft.core.stores.WorldStore;
import work.xeltica.craft.core.utils.DiscordService;

/**
 * ワールド制御に関するハンドラーをまとめています。
 * TODO: 機能別に再編
 * @author Xeltica
 */
public class WorldHandler implements Listener {
    public WorldHandler() {
    }

    @EventHandler
    /**
     * 進捗を達成できるワールドを限定させるハンドラー
     */
    public void onAdvancementDone(PlayerAdvancementDoneEvent e) {
        final var p = e.getPlayer();
        if (!advancementWhitelist.contains(p.getWorld().getName())) {
            final var advancement = e.getAdvancement();

            for (var criteria : advancement.getCriteria()) {
                p.getAdvancementProgress(advancement).revokeCriteria(criteria);
            }
        }
    }

    @EventHandler
    /**
     * サンドボックスでのエンダーチェスト設置を防止する
     */
    public void onBlockPlace(BlockPlaceEvent e) {
        final var p = e.getPlayer();
        if (p.getWorld().getName().equals("sandbox")) {
            final var block = e.getBlock().getType();
            // エンダーチェストはダメ
            if (block == Material.ENDER_CHEST) {
                e.setCancelled(true);
            }
        }
    }

    @EventHandler
    /**
     * テレポート時のいろいろなガード機能など
     * TODO: 分割する
     */
    public void onPlayerTeleportGuard(PlayerTeleportEvent e) {
        final var p = e.getPlayer();
        final var world = e.getTo().getWorld();
        final var name = world.getName();
        final var store = WorldStore.getInstance();

        final var isLockedWorld = store.isLockedWorld(name);
        final var isCreativeWorld = store.isCreativeWorld(name);
        final var displayName = store.getWorldDisplayName(name);
        final var desc = store.getWorldDescription(name);

        final var from = e.getFrom().getWorld();
        final var to = e.getTo().getWorld();

        final var fromId = from.getUID();
        final var toId = to.getUID();

        if (fromId.equals(toId)) return;

        if (isLockedWorld && !p.hasPermission("hub.teleport." + name)) {
            p.playSound(p.getLocation(), Sound.BLOCK_ANVIL_PLACE, 1, 0.5f);
            p.sendMessage("§aわかば§rプレイヤーは§6" + displayName + "§rに行くことができません！\n§b/promo§rコマンドを実行して、昇格方法を確認してください！");
            e.setCancelled(true);
            return;
        }

        WorldStore.getInstance().saveCurrentLocation(p);

        final var hint = switch (to.getName()) {
            case "main" -> Hint.GOTO_MAIN;
            case "hub2" -> Hint.GOTO_LOBBY;
            case "wildarea2" -> Hint.GOTO_WILDAREA;
            case "wildarea2_nether" -> Hint.GOTO_WILDNETHER;
            case "wildarea2_the_end" -> Hint.GOTO_WILDEND;
            case "wildareab" -> Hint.GOTO_WILDAREAB;
            case "sandbox2" -> Hint.GOTO_SANDBOX;
            case "art" -> Hint.GOTO_ART;
            case "nightmare2" -> Hint.GOTO_NIGHTMARE;
            case "shigen_nether" -> Hint.GOTO_WILDNETHERB;
            case "shigen_end" -> Hint.GOTO_WILDENDB;
            default -> null;
        };

        // 以前サーバーに来ている or FIRST_SPAWN フラグが立っている
        final var isNotFirstTeleport = p.hasPlayedBefore() || PlayerStore.getInstance().open(p).getBoolean(PlayerDataKey.FIRST_SPAWN);

        if (to.getName().equals("main") && isNotFirstTeleport && !HintModule.INSTANCE.hasAchieved(p, Hint.GOTO_MAIN)) {
            // はじめてメインワールドに入った場合、対象のスタッフに通知する
            try {
                DiscordService.getInstance().alertNewcomer(p);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }

        if (hint != null && isNotFirstTeleport) {
            HintModule.INSTANCE.achieve(p, hint);
        }

        if (isCreativeWorld) {
            p.setGameMode(GameMode.CREATIVE);
        }
        if (name.equals("nightmare2")) {
            world.setDifficulty(Difficulty.HARD);
            world.setGameRule(GameRule.DO_DAYLIGHT_CYCLE, false);
            world.setGameRule(GameRule.DO_WEATHER_CYCLE, false);
            world.setGameRule(GameRule.MOB_GRIEFING, false);
            world.setTime(18000);
            world.setStorm(true);
            world.setWeatherDuration(20000);
            world.setThundering(true);
            world.setThunderDuration(20000);
            p.playSound(p.getLocation(), Sound.BLOCK_END_PORTAL_SPAWN, SoundCategory.PLAYERS, 1, 0.5f);
        }
        if (desc != null) {
            p.sendMessage(desc);
        }

        Bukkit.getScheduler().runTaskLater(XCorePlugin.getInstance(), () -> {
            PlayerStore.getInstance().open(p).set(PlayerDataKey.FIRST_SPAWN, true);
        }, 20 * 5);
    }

    @EventHandler
    /**
     * ワールドを移動した時に、ワールド名を表示する機能
     */
    public void onPlayerMoveWorld(PlayerChangedWorldEvent e) {
        final var name = WorldStore.getInstance().getWorldDisplayName(e.getPlayer().getWorld());
        e.getPlayer().showTitle(Title.title(Component.text(name).color(TextColor.color(0xFFB300)), Component.empty()));
    }

    @EventHandler
    /**
     * 自動クラフト機能
     */
    public void onDispenser(BlockPreDispenseEvent event) {
        final var block = event.getBlock();
        if (block.getBlockData().getMaterial() != Material.DISPENSER) return;
        GlowItemFrame itemFlame = null;

        for (Entity entity: block.getLocation().toCenterLocation().getNearbyEntitiesByType(GlowItemFrame.class, 1)) {
            if (entity instanceof GlowItemFrame flame) {
                if (flame.getLocation().add(flame.getAttachedFace().getDirection()).getBlock().getLocation().toBlockLocation().equals(block.getLocation().toBlockLocation())) {
                    itemFlame = flame;
                    break;
                }
            }
        }
        if (itemFlame == null) return;
        event.setCancelled(true);

        final var item = itemFlame.getItem();
        final var recipes = new ArrayList<CraftRecipe>();
        for (Recipe recipe: Bukkit.getServer().getRecipesFor(item)) {
            if (recipe instanceof ShapedRecipe shapedRecipe) {
                final var ingredients = new ArrayList<ItemStack>();
                for (ItemStack i: shapedRecipe.getIngredientMap().values()) {
                    if (i == null) continue;
                    ingredients.add(new ItemStack(i.getType(),1));
                }
                recipes.add(new CraftRecipe(ingredients, recipe.getResult()));
            }
            if (recipe instanceof ShapelessRecipe shapelessRecipe) {
                final var ingredients = new ArrayList<ItemStack>();
                for (ItemStack i: shapelessRecipe.getIngredientList()) {
                    if (i == null) continue;
                    ingredients.add(new ItemStack(i.getType(), 1));
                }
                recipes.add(new CraftRecipe(ingredients, recipe.getResult()));
            }
        }

        if (block.getState() instanceof Dispenser dispenser) {
            final var inventory = dispenser.getInventory();

            for (CraftRecipe recipe: recipes) {
                final var fixedInventory = Arrays.stream(inventory.getContents()).filter(Objects::nonNull).collect(
                        Collectors.groupingBy(ItemStack::getType, Collectors.summingInt(ItemStack::getAmount))
                );
                var flag = true;
                for (Material ingredient: recipe.getFixedRecipe().keySet()) {
                    if (fixedInventory.containsKey(ingredient) && fixedInventory.get(ingredient) >= recipe.getFixedRecipe().get(ingredient)) {
                        fixedInventory.replace(ingredient, fixedInventory.get(ingredient) - recipe.getFixedRecipe().get(ingredient));
                        continue;
                    }
                    flag = false;
                }

                if (flag) {
                    for (ItemStack itemStack: inventory.getContents()) {
                        if (itemStack == null) continue;
                        inventory.remove(itemStack);
                    }
                    for (Map.Entry<Material, Integer> entry: fixedInventory.entrySet()) {
                        final var itemStack = new ItemStack(entry.getKey());
                        itemStack.setAmount(entry.getValue());
                        inventory.addItem(itemStack);
                    }

                    if (event.getBlock().getState().getData() instanceof Directional directional) {
                        final var location = block.getLocation().toBlockLocation().add(directional.getFacing().getDirection());
                        if (location.getBlock().getState() instanceof BlockInventoryHolder inventoryHolder) {
                            final var result = inventoryHolder.getInventory().addItem(recipe.result());
                            if (!result.isEmpty()) {
                                for (ItemStack i: result.values()) {
                                    block.getWorld().dropItem(location, i);
                                }
                            }
                            return;
                        }
                        block.getWorld().dropItem(location, recipe.result());
                        return;
                    }
                }
            }
        }
    }

    private final Set<String> advancementWhitelist = new HashSet<>(List.of(
        "wildarea2",
        "wildarea2_nether",
        "wildarea2_the_end",
        "wildareab",
        "shigen_nether",
        "shigen_end",
        "main",
        "nightmare2"
    ));
}

File: src/main/java/work/xeltica/craft/core/handlers/WorldHandler.kt
package work.xeltica.craft.core.handlers

import io.papermc.paper.event.block.BlockPreDispenseEvent
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.title.Title
import org.bukkit.*
import org.bukkit.block.Dispenser
import org.bukkit.entity.GlowItemFrame
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.player.PlayerAdvancementDoneEvent
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.event.player.PlayerTeleportEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.inventory.ShapelessRecipe
import org.bukkit.material.Directional
import work.xeltica.craft.core.XCorePlugin.Companion.instance
import work.xeltica.craft.core.api.playerStore.PlayerStore
import work.xeltica.craft.core.hooks.DiscordHook
import work.xeltica.craft.core.models.CraftRecipe
import work.xeltica.craft.core.modules.hint.Hint
import work.xeltica.craft.core.modules.hint.HintModule.achieve
import work.xeltica.craft.core.modules.hint.HintModule.hasAchieved
import work.xeltica.craft.core.modules.player.PlayerDataKey
import work.xeltica.craft.core.modules.world.WorldModule
import work.xeltica.craft.core.modules.world.WorldModule.getWorldDisplayName

/**
 * ワールド制御に関するハンドラーをまとめています。
 * TODO: 機能別に再編
 * @author Xeltica
 */
class WorldHandler : Listener {
    /*
     * 進捗を達成できるワールドを限定させるハンドラー
     */
    @EventHandler
    fun onAdvancementDone(e: PlayerAdvancementDoneEvent) {
        val p = e.player
        if (!advancementWhitelist.contains(p.world.name)) {
            val advancement = e.advancement
            for (criteria in advancement.criteria) {
                p.getAdvancementProgress(advancement).revokeCriteria(criteria!!)
            }
        }
    }

    /*
     * サンドボックスでのエンダーチェスト設置を防止する
     */
    @EventHandler
    fun onBlockPlace(e: BlockPlaceEvent) {
        val p = e.player
        if (p.world.name == "sandbox") {
            val block = e.block.type
            // エンダーチェストはダメ
            if (block == Material.ENDER_CHEST) {
                e.isCancelled = true
            }
        }
    }

    /*
     * テレポート時のいろいろなガード機能など
     * TODO: 分割する
     */
    @EventHandler
    fun onPlayerTeleportGuard(e: PlayerTeleportEvent) {
        val p = e.player
        val world = e.to.world
        val name = world.name
        val worldModule = WorldModule
        val isLockedWorld = worldModule.isLockedWorld(name)
        val isCreativeWorld = worldModule.isCreativeWorld(name)
        val displayName = worldModule.getWorldDisplayName(name)
        val desc = worldModule.getWorldDescription(name)
        val from = e.from.world
        val to = e.to.world
        val fromId = from.uid
        val toId = to.uid
        if (fromId == toId) return
        if (isLockedWorld && !p.hasPermission("hub.teleport.$name")) {
            p.playSound(p.location, Sound.BLOCK_ANVIL_PLACE, 1f, 0.5f)
            p.sendMessage("§aわかば§rプレイヤーは§6$displayName§rに行くことができません！\n§b/promo§rコマンドを実行して、昇格方法を確認してください！")
            e.isCancelled = true
            return
        }
        worldModule.saveCurrentLocation(p)
        val hint = when (to.name) {
            "main" -> Hint.GOTO_MAIN
            "hub2" -> Hint.GOTO_LOBBY
            "wildarea2" -> Hint.GOTO_WILDAREA
            "wildarea2_nether" -> Hint.GOTO_WILDNETHER
            "wildarea2_the_end" -> Hint.GOTO_WILDEND
            "wildareab" -> Hint.GOTO_WILDAREAB
            "sandbox2" -> Hint.GOTO_SANDBOX
            "art" -> Hint.GOTO_ART
            "nightmare2" -> Hint.GOTO_NIGHTMARE
            "shigen_nether" -> Hint.GOTO_WILDNETHERB
            "shigen_end" -> Hint.GOTO_WILDENDB
            else -> null
        }

        // 以前サーバーに来ている or FIRST_SPAWN フラグが立っている
        val isNotFirstTeleport = p.hasPlayedBefore() || PlayerStore.open(p).getBoolean(PlayerDataKey.FIRST_SPAWN)
        if (to.name == "main" && !hasAchieved(p, Hint.GOTO_MAIN)) {
            // はじめてメインワールドに入った場合、対象のスタッフに通知する
            try {
                DiscordHook.alertNewcomer(p)
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
        if (hint != null && isNotFirstTeleport) {
            achieve(p, hint)
        }
        if (isCreativeWorld) {
            p.gameMode = GameMode.CREATIVE
        }
        if (name == "nightmare2") {
            world.difficulty = Difficulty.HARD
            world.setGameRule(GameRule.DO_DAYLIGHT_CYCLE, false)
            world.setGameRule(GameRule.DO_WEATHER_CYCLE, false)
            world.setGameRule(GameRule.MOB_GRIEFING, false)
            world.time = 18000
            world.setStorm(true)
            world.weatherDuration = 20000
            world.isThundering = true
            world.thunderDuration = 20000
            p.playSound(p.location, Sound.BLOCK_END_PORTAL_SPAWN, SoundCategory.PLAYERS, 1f, 0.5f)
        }
        if (desc != null) {
            p.sendMessage(desc)
        }
        Bukkit.getScheduler()
            .runTaskLater(instance, Runnable { PlayerStore.open(p)[PlayerDataKey.FIRST_SPAWN] = true }, 20 * 5)
    }

    /**
     * ワールドを移動した時に、ワールド名を表示する機能
     */
    @EventHandler
    fun onPlayerMoveWorld(e: PlayerChangedWorldEvent) {
        val name = getWorldDisplayName(e.player.world)
        e.player.showTitle(Title.title(Component.text(name).color(TextColor.color(0xFFB300)), Component.empty()))
    }

    /*
     * 自動クラフト機能
     */
    @EventHandler
    fun onDispenser(event: BlockPreDispenseEvent) {
        val block = event.block
        val blockData = block.blockData
        if (blockData.material != Material.DISPENSER) return
        val frames = block.location.toCenterLocation().getNearbyEntitiesByType(GlowItemFrame::class.java, 1.0)
        val itemFrame = frames.firstOrNull {
            it.location.clone()
                .add(it.attachedFace.direction).block.location.toBlockLocation() == block.location.toBlockLocation()
        } ?: return

        event.isCancelled = true
        val item = itemFrame.item

        val recipes = ArrayList<CraftRecipe>()
        for (recipe in Bukkit.getServer().getRecipesFor(item)) {
            if (recipe is ShapedRecipe) {
                val ingredients = ArrayList<ItemStack>()
                for (i in recipe.ingredientMap.values) {
                    if (i == null) continue
                    ingredients.add(ItemStack(i.type, 1))
                }
                recipes.add(CraftRecipe(ingredients, recipe.getResult()))
            }
            if (recipe is ShapelessRecipe) {
                val ingredients = ArrayList<ItemStack>()
                for (i in recipe.ingredientList) {
                    if (i == null) continue
                    ingredients.add(ItemStack(i.type, 1))
                }
                recipes.add(CraftRecipe(ingredients, recipe.getResult()))
            }
        }
        val state = block.state
        if (state is Dispenser) {
            val inventory = state.inventory
            val contents = inventory.contents.filterNotNull()

            for (recipe in recipes) {
                val fixedInventory = contents
                    .groupingBy { it.type }
                    .fold(0) { acc, i -> acc + i.amount }
                    .toMutableMap()
                var flag = true
                for (ingredient in recipe.fixedRecipe.keys) {
                    if (fixedInventory.containsKey(ingredient) && fixedInventory[ingredient]!! >= recipe.fixedRecipe[ingredient]!!) {
                        fixedInventory.replace(
                            ingredient,
                            fixedInventory[ingredient]!! - recipe.fixedRecipe[ingredient]!!
                        )
                        continue
                    }
                    flag = false
                }
                if (flag) {
                    for (itemStack in contents) {
                        inventory.remove(itemStack)
                    }
                    for ((key, value) in fixedInventory) {
                        val itemStack = ItemStack(key)
                        itemStack.amount = value
                        inventory.addItem(itemStack)
                    }
                    if (blockData is Directional) {
                        val location: Location = block.location.toBlockLocation().add(blockData.facing.direction)
                        val result: HashMap<Int, ItemStack> = state.inventory.addItem(recipe.result)
                        if (result.isNotEmpty()) {
                            for (i in result.values) {
                                block.world.dropItem(location, i)
                            }
                        }
                        return
                    }
                }
            }
        }
    }

    private val advancementWhitelist: Set<String> = setOf(
        "wildarea2",
        "wildarea2_nether",
        "wildarea2_the_end",
        "wildareab",
        "shigen_nether",
        "shigen_end",
        "main",
        "nightmare2"
    )
}